Fix worker taints not applied

orangedeng opened the pull request:
**Problem:**
Setting node pool taints and node template taints doesn't work.

**Solution:**
Add kubelet args `--register-with-taints` parameters to kubelet worker
process. So those taints will be added after kubelet is up and running.

Related issue:
https://github.com/rancher/rancher/issues/22513

gitlawr commented on pkg/controllers/management/node/controller.go:
```diff
@@ -537,7 +537,7 @@ func (m *Lifecycle) saveConfig(config *nodeconfig.NodeConfig, nodeDir string, ob
 		// the taints from node pool should override the taints from template.
 		if _, ok := nodeSet[key]; !ok {
-			expectTaints = append(obj.Spec.DesiredNodeTaints, template.Spec.NodeTaints[ti])
+			expectTaints = append(expectTaints, template.Spec.NodeTaints[ti])
```
Can you explain this change

orangedeng replied:
This is a bug. We should append `expectTaints` list instead of `obj.Spec.DesiredNodeTaints`.
We need to combine taints from node template and node pool.

gitlawr commented on pkg/rkenodeconfigserver/nodeserver.go:
```diff
@@ -201,7 +202,7 @@ func (n *RKENodeConfigServer) nodeConfig(ctx context.Context, cluster *v3.Cluste
 				nc.Processes = enhanceWindowsProcesses(tempNode.Processes)
```
Is the kubelet param needed by windows workers?

orangedeng replied:
Updated.

orangedeng commented:
Updated

gitlawr commented:
Also, would like to know the reason for using this kubelet argument. If the taints are not added on registration, won't the framework add them in a later time?

orangedeng commented:
Unlike rke cli, rancher cluster provisioner only cares about control plane and etcd node and execute rke up after the configuration of those nodes is updated. That’s what we need to modify the rke worker to add taints to kubelet parameter.

alena1108 commented on pkg/rkenodeconfigserver/nodeserver.go:
```diff
@@ -291,3 +294,12 @@ func enhanceWindowsProcesses(processes map[string]v3.Process) map[string]v3.Proc
+
+func appendTaintsToKubeletArgs(processes map[string]v3.Process, initialTaints []v3.RKETaint) map[string]v3.Process {
+	if kubelet, ok := processes["kubelet"]; ok && len(initialTaints) != 0 {
+		taintArgs := fmt.Sprintf("--register-with-taints=%s", strings.Join(taints.GetStringsFromRKETaint(initialTaints), ","))
```
@superseb can you think of any implications of using this flag on a kubelet start

alena1108 replied:
@orangedeng would - and should - user be able to override this flag via passing custom kubelet arguments in cluster.yml config

superseb replied:
We use the same on controlplane nodes, as long as it can take updates from the Rancher side (override), it should be fine

orangedeng replied:
As kubelet can only accept only one `--register-with-taints` parameter, I think we should merge the taints from user custom taints and our taints.

orangedeng commented:
Updated PR.
1. Added tests for code
2. We will merge those taints in kubelet `extra_args` and node pool/template taints. And taints from pool/template will override the taints from kubelet `extra_args`.

orangedeng commented:
Addressed comments.
